src/main/java/com/google/devtools/build/lib/cmdline/IgnoredSubdirectories.java

```diff
@@ -14,57 +14,123 @@
+    public void serialize(
+        SerializationContext context, IgnoredSubdirectories obj, CodedOutputStream codedOut)
+        throws SerializationException, IOException {
+      context.serialize(obj.prefixes, codedOut);
+      context.serialize(obj.patterns, codedOut);
     }
+
+    @Override
+    public IgnoredSubdirectories deserialize(
+        DeserializationContext context, CodedInputStream codedIn)
+        throws SerializationException, IOException {
+      ImmutableSet<PathFragment> prefixes = context.deserialize(codedIn);
+      ImmutableList<String> patterns = context.deserialize(codedIn);
+
+      return new IgnoredSubdirectories(prefixes, patterns);
+    }
+  }
+
+  private IgnoredSubdirectories(
+      ImmutableSet<PathFragment> prefixes, ImmutableList<String> patterns) {
     this.prefixes = prefixes;
+    this.patterns = patterns;
+    this.splitPatterns =
+        patterns.stream()
+            .map(p -> Iterables.toArray(SLASH_SPLITTER.split(p), String.class))
+            .collect(toImmutableList());
   }
 
   public static IgnoredSubdirectories of(ImmutableSet<PathFragment> prefixes) {
-    if (prefixes.isEmpty()) {
+    return of(prefixes, ImmutableList.of());
+  }
+
+  public static IgnoredSubdirectories of(
+      ImmutableSet<PathFragment> prefixes, ImmutableList<String> patterns) {
+    if (prefixes.isEmpty() && patterns.isEmpty()) {
       return EMPTY;
-    } else {
-      return new IgnoredSubdirectories(prefixes);
     }
+
+    for (PathFragment prefix : prefixes) {
+      Preconditions.checkArgument(!prefix.isAbsolute());
+    }
+
+    return new IgnoredSubdirectories(prefixes, patterns);
   }
 
   public IgnoredSubdirectories withPrefix(PathFragment prefix) {
-    ImmutableSet<PathFragment> prefixed =
+    Preconditions.checkArgument(!prefix.isAbsolute());
+
+    ImmutableSet<PathFragment> prefixedPrefixes =
         prefixes.stream().map(prefix::getRelative).collect(toImmutableSet());
-    return new IgnoredSubdirectories(prefixed);
+
+    ImmutableList<String> prefixedPatterns =
+        patterns.stream().map(p -> prefix + "/" + p).collect(toImmutableList());
+
+    return new IgnoredSubdirectories(prefixedPrefixes, prefixedPatterns);
   }
 
   public IgnoredSubdirectories union(IgnoredSubdirectories other) {
     return new IgnoredSubdirectories(
-        ImmutableSet.<PathFragment>builder().addAll(prefixes).addAll(other.prefixes).build());
+        ImmutableSet.<PathFragment>builder().addAll(prefixes).addAll(other.prefixes).build(),
+        ImmutableList.<String>builder().addAll(patterns).addAll(other.patterns).build());
   }
 
   /** Filters out entries that cannot match anything under {@code directory}. */
   public IgnoredSubdirectories filterForDirectory(PathFragment directory) {
     ImmutableSet<PathFragment> filteredPrefixes =
         prefixes.stream().filter(p -> p.startsWith(directory)).collect(toImmutableSet());
+    ImmutableList.Builder<String> filteredPatterns = ImmutableList.builder();
```

Comment:
were you planning to do something here? if not, just use ImmutableList.of() below

Reply:
Argh, that's embarrassing. I thought I wrote a test battery for `IgnoredSubdirectories`, but I either did not or managed to lose it somewhere. Anyways, I wrote one now and fixed this embarrassing mistake.